spices/SPICE-0008-member-deletion.adoc

```diff
@@ -0,0 +1,345 @@
+
+[source,pkl]
+----
+prop {
+  foo = "baz"
+  bar = false
+  ["bar"] = "baz"
+  28
+  quux = false
+  ["qux"] = false
+  496
+}
+----
+
+Amending existing members or adding new ones is straightforward.
+Deleting members, however, is not.
+If, in the above example, the entry `"baz"` should be deleted, there is no "direct" way of doing so.
+Instead, `prop` would first need to be converted to a `Map`, then `"baz"` would be removed, and finally the resulting `Map` would be converted back to a `Dynamic`;
+
+[source,pkl]
+----
+amends "base.pkl"
+
+prop = (super.prop.toMap().remove("foo").toDynamic()) {
+  [0] = super[0] - 14 // <1>
+  496
+  ["qux"] = bar
+  quux = false
+}
+----
+<1> No longer amends anything (see below)
+
+Evaluating this adjusted `amending.pkl` renders the following result.
+
+[source,pkl]
+----
+prop { // <1>
+  bar = "baz" // <2>
+  quux = false
+  496
+  ["qux"] = "baz" // <2>
+}
+----
+<1> There is no more element `28`.
+<2> Entry `"bar"` has been transformed into property `bar`, the original `bar` property has been discarded, and entry `"qux"` now refers to the new `bar` property.
+
+Besides being cumbersome, this method has multiple additional issues:
+
+1. When a `Dynamic` has a property with the same name as a `String`-keyed entry, `toMap()` retains the entry and drops the property.
+2. Regardless of overlap, when transforming back `toDynamic()`, all members that were previously _properties_ or _entries_ now become _properties_.
+3. Members defined lazily, in terms of other members, lose this late-bound relationship, because `toMap()` forces the evaluation of all properties and entries to _values_ (instead of late-bound expressions).
+4. Changing to and from a `Map` discards the existing order of the object members; the _properties_ in `super.prop` are grouped at the start, followed by the _entries_ of `super.prop`.
+5. This approach discards all _elements_ in `super.prop`.
+
+That last point about the loss of elements is not easily worked around and has some further consequences.
+The object member `[0] = super[0] - 14` no longer has an existing element to amend, so it is ignored entirely.
+Explicitly spreading the elements of `super.prop` into this new `Dynamic` would cause this expression to trigger a ``Duplicate definition of member `0``` error.
+
+``Listing``s and ``Mapping``s only have _elements_, respectively, _entries_, whose order is preserved when transformed `toList()`, respectively, `toMap()`, so they only suffer from issue 3 and the cumbersome formulation.
+
+== Proposed Solution
+
+A new keyword, `delete`, is introduced to the language.
+This keyword can be used in the expression position of an assignment-style object member specification.
+
+With this new keyword, `amending.pkl` from the <<Motivation>> can be reformulated as follows.
+
+[source,pkl]
+----
+amends "base.pkl"
+
+prop {
+  [0] = super[0] - 14
+  496
+  ["quux"] = foo
+  ["baz"] = delete // <1>
```

Comment:
Have you consider an alternate syntax for `delete`?
This seems very similar to normal assigning syntax and, specially without syntax highlighting, may fly under the radar while reviewing code.

What about

```pkl
prop {
  delete ["entry"] // for entries
  delete 0 // for indexes
  delete prop // for properties
  delete [[foo == 10]] // for predicates
}
```

It's easier to see the change and totally unambiguous for the parser.

Reply:
I had considered this, but opted against it for a few reasons (that I'll include in `Alternatives considered`); I like that in your change, deletes are emphasised over other amends/overwrites. The downsides, however:
* currently and with my proposed syntax, you can skim the left-margin of an amends expression to find all things that are affected
* your changes requires member predicates and entries to occur in different places (namely in the argument position of `delete`) than they do now
* `delete prop`, to me, reads like an expression that would be evaluated as usual, i.e. resolving and evaluating `prop` to use its _value_ rather than its _name_, which is particularly painful considering your index suggestion (without square brackets); surely that could be done parametrically ;)
* as shown here, such alternative syntax makes room for all sorts of small variations (here, that element indices don't have square brackets) that allow subtle bugs (we've been talking about ways to disambiguate entry-indices from integer-valued entry keys), and require more syntax knowledge of the user.

Reply:
I actually like what @stackoverflow 's proposal better.

> * currently and with my proposed syntax, you can skim the left-margin of an amends expression to find all things that are affected

Is this just for readability, or has any impact to the implementation? This feels like a nice to have feature. But `= delete` feels odd semantically. Most people would think the right side of `=` to be a value instead of an operator.

> * `delete prop`, to me, reads like an expression that would be evaluated as usual, i.e. resolving and evaluating `prop` to use its _value_ rather than its _name_, which is particularly painful considering your index suggestion (without square brackets); surely that could be done parametrically ;)

Would `delete [0]` address the concern? It feels reasonable to add brackets around indices.